### PR TITLE
[VDO-5672] Ensure the bio bi_ioprio property is respected and update the internal fake/linux bio implementation

### PR DESCRIPTION
**Pull request by slegendr**
VDO already handles the bio->bi_ioprio property correctly in most places. Only a small modification in vio.c was needed to explicitly reset the property when the bio is reset. (first commit)

However, the internal fake/linux bio implementation needed updating to align it with the upstream 6.10 kernel. This included removing unused files/directories (second commit), and general cleanup to update the data types and function declarations/definitions, along with removing kernel functions declared but unused by the VDO codebase (third commit).

Additional cleanup was also performed to remove USE_ALTERNATE preprocessor directives that are no longer needed from the base code. (fourth commit)

Compilation and basic checkin testing against the 6.11 and 6.12 kernels in our lab has also been completed successfully, indicating no additional changes are needed.

### CODE REVIEW

File: src/c++/uds/userLinux/uds/asm/unaligned.h [hunk 1]
```diff
@@ -10,35 +10,48 @@
...
-#define COMPARE(x, y, op)                              \
-	__builtin_choose_expr(                         \
-		(TYPECHECK(x, y) && CONSTCHECK(x, y)), \
-		(((x) op(y)) ? (x) : (y)),             \
-		SAFE_COMPARE(x, y, UNIQUE_ID(x_), UNIQUE_ID(y_), op))
+#define __careful_cmp(op, x, y)                            \
+	__builtin_choose_expr(                             \
+		(__typecheck(x, y) && __constcheck(x, y)), \
+		__cmp(op, x, y),		           \
+		__cmp_once(op, x, y, __UNIQUE_ID(x_), __UNIQUE_ID(y_)))
```
Comment (lorelei-sakai) [inline src/c++/uds/userLinux/uds/asm/unaligned.h:51]: It's a little hard to tell from the diffs, but I think there's a whitespace issue here. Between the visible text and the line continuation character ("\\"), there should be only spaces, no tabs. I think there are some tabs in there.

EDIT: Ugh, reading through the rest of the files, kernel seem to like using tabs for this kind of thing. So either spaces or tabs is fine, but all lines should at least be consistent with each other in this macro.

File: src/c++/uds/userLinux/uds/linux/blkdev.h [hunk 2]
```diff
@@ -38,20 +47,44 @@ struct block_device {
...
+	int         error;
+	const char *name;
+} blk_errors[] = {
+	[BLK_STS_OK]           = { 0,		"" },
+	[BLK_STS_NOSPC]        = { -ENOSPC,	"critical space allocation" },
+	[BLK_STS_RESOURCE]     = { -ENOMEM,	"kernel resource" },
+	
+	/* error specifically for VDO unit tests */
+	[BLK_STS_VDO_INJECTED] = { 31,         "vdo injected error" },
+	/* everything else not covered above: */
```
Comment (lorelei-sakai) [inline src/c++/uds/userLinux/uds/linux/blkdev.h:61]: Another whitespace difference. Most of these lines seem to use tabs between the errno and the string, so line 60 should match. (I don't really like having tabs there, but that's the consistent thing to do.)

File: src/c++/vdo/fake/asm/unaligned.h [hunk 3]
```diff
@@ -1,17 +1,12 @@
-/*
- * FOR INTERNAL USE ONLY, DO NOT DISTRIBUTE!!!!
+/* SPDX-License-Identifier: GPL-2.0-only */
  *
  * Unit test implementation of asm/unaligned.h
  *
```
Comment (lorelei-sakai) [inline src/c++/vdo/fake/asm/unaligned.h:4]: Huh. So this file confused me because it should conflict with the uds asm/unaligned.h header. But it looks like nothing in VDO uses the two macros defined here, so we could probably remove the file entirely. Since that likely entails some manifest changes, etc., I'd make a separate commit before this one to remove the file and adjust anything else that needs adjusting.

File: src/c++/vdo/fake/linux/bio.h [hunk 4]
```diff
@@ -323,212 +139,61 @@ struct bio_integrity_payload {
...
 #ifdef CONFIG_BLK_CGROUP
 void bio_associate_blkg(struct bio *bio);
-void bio_associate_blkg_from_css(struct bio *bio,
-				 struct cgroup_subsys_state *css);
-void bio_clone_blkg_association(struct bio *dst, struct bio *src);
+//void bio_associate_blkg_from_css(struct bio *bio,
+//				 struct cgroup_subsys_state *css);
+//void bio_clone_blkg_association(struct bio *dst, struct bio *src);
+//void blkcg_punt_bio_submit(struct bio *bio);
 #else	/* CONFIG_BLK_CGROUP */
```
Comment (lorelei-sakai) [inline src/c++/vdo/fake/linux/bio.h:175]: Why are these lines commented out? If we're not using them, they should just be removed.
Ditto for the block of commented lines below this, as well.

File: src/c++/vdo/fake/linux/blk_types.h [hunk 5]
```diff
@@ -15,9 +15,22 @@
...
+ */
+#define BLK_STS_OK 0
+#define BLK_STS_NOSPC           ((blk_status_t)3)
+#define BLK_STS_RESOURCE        ((blk_status_t)9)
+#define BLK_STS_IOERR           ((blk_status_t)10)
+
+/* hack for VDO, don't use elsewhere */
+#define BLK_STS_VDO_INJECTED    ((blk_status_t)31)
+
 /*
```
Comment (lorelei-sakai) [inline src/c++/vdo/fake/linux/blk_types.h:34]: Why do we need to duplicate this block? Can't this just include the uds blkdev.h header?

File: src/c++/vdo/fake/linux/blk_types.h [hunk 6]
```diff
@@ -151,65 +170,49 @@ enum req_opf {
...
-	REQ_OP_ZONE_RESET_ALL	= 17,
+	REQ_OP_ZONE_RESET_ALL	= 15,
 
 	/* Driver private requests */
 	REQ_OP_DRV_IN		= 34,
 	REQ_OP_DRV_OUT		= 35,
 
-	REQ_OP_LAST,
+	REQ_OP_LAST             = 36,
 };
```
Comment (lorelei-sakai) [inline src/c++/vdo/fake/linux/blk_types.h:183]: All previous lines seem to have tabs before the '=', could this line match?

File: src/c++/vdo/fake/linux/highmem.h [hunk 7]
```diff
@@ -1,8 +1,8 @@
-/* SPDX-License-Identifier: GPL-2.0-only */
+/* SPDX-License-Identifier: GPL-2.0 */
 /*
```
Comment (lorelei-sakai) [inline src/c++/vdo/fake/linux/highmem.h:2]: Must other files have license GPL-2.0-only. I think removing the "-only" here was a mistake? (Not that licensing on these matters all that much...)

File: src/c++/vdo/fake/linux/timer.h [hunk 8]
```diff
@@ -1,16 +1,21 @@
 /* SPDX-License-Identifier: GPL-2.0-only */
 /*
- * Copyright 2023 Red Hat
- *
- * $Id$
+ * Unit test implementations of linux/timer.h.
+ *
```
Comment (lorelei-sakai) [inline src/c++/vdo/fake/linux/timer.h:4]: Trailing space on this line is unnecessary.

File: src/c++/vdo/fake/linux/wait.h [hunk 9]
```diff
@@ -14,15 +14,16 @@
 
 #include "thread-utils.h"
 
+/* Fields which are not used in VDO unit tests are excluded. */
 typedef struct wait_queue_head {
-	spinlock_t		lock;
+	spinlock_t          	lock;
 	struct list_head	head;
```
Comment (lorelei-sakai) [inline src/c++/vdo/fake/linux/wait.h:20]: This line and the later one (23/24) change whitespace for no reason; could you change it back? Diffs are long enough as it is.

**Review by lorelei-sakai — CHANGES_REQUESTED** [review]
I think my comments are mostly fiddly things related to the fake files. None of that is too important so if any of the comments are hard to deal with, just ignore them. I think only the third commit will go upstream? In which case, it all generally looks good.

File: src/c++/uds/userLinux/uds/asm/unaligned.h (hunk 1 above)
Comment (slegendr) [reply src/c++/uds/userLinux/uds/asm/unaligned.h:51]: I verified that it's all spaces between the expressions and the continuation character. I have no idea why it's displaying like this in the diff.

File: src/c++/uds/userLinux/uds/linux/blkdev.h (hunk 2 above)
Comment (slegendr) [reply src/c++/uds/userLinux/uds/linux/blkdev.h:61]: Fixed.

File: src/c++/vdo/fake/asm/unaligned.h (hunk 3 above)
Comment (slegendr) [reply src/c++/vdo/fake/asm/unaligned.h]: Added a commit to remove this. Our MANIFEST only specifies the c++/uds/userLinux/uds/asm/unaligned.h, so no changes were needed.

File: src/c++/vdo/fake/linux/blk_types.h (hunk 6 above)
Comment (slegendr) [reply src/c++/vdo/fake/linux/blk_types.h:172]: Fixed.

File: src/c++/vdo/fake/linux/timer.h [hunk 10]
```diff
@@ -1,16 +1,21 @@
...
  */
 
 #ifndef LINUX_TIMER_H
 #define LINUX_TIMER_H
 
-#include <linux/jiffies.h>
 #include <linux/list.h>
 
+/* Defined in linux/timer_types.h
+ *
```
Comment (lorelei-sakai) [inline src/c++/vdo/fake/linux/timer.h:14]: Line 14 still has a trailing space after the *

File: src/c++/vdo/fake/linux/wait.h [hunk 11]
```diff
@@ -14,15 +14,16 @@
...
+/* Fields which are not used in VDO unit tests are excluded. */
 typedef struct wait_queue_head {
 	spinlock_t		lock;
 	struct list_head	head;
 } wait_queue_head_t;
 
-struct wait_queue_entry {
-	void		       *private;
+typedef struct wait_queue_entry {
+	void                    *private;
```
Comment (lorelei-sakai) [inline src/c++/vdo/fake/linux/wait.h:24]: I think this line is misaligned. the 'p' or "private" should line up with the 'e' of "entry".


Replacing tabs with spaces is good, though.

**Review by lorelei-sakai — CHANGES_REQUESTED** [review]
Just a few really minor formatting issues, otherwise looks good.

File: src/c++/vdo/fake/linux/timer.h (hunk 10 above)
Comment (slegendr) [reply src/c++/vdo/fake/linux/timer.h]: Fixed.

File: src/c++/vdo/fake/linux/wait.h (hunk 11 above)
Comment (slegendr) [reply src/c++/vdo/fake/linux/wait.h]: Fixed.

**Review by lorelei-sakai — APPROVED** [review]
Looks good.
